02_knight_and_princess/roman_kuprii/main.cpp

```diff
@@ -0,0 +1,131 @@
+#include <array>
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+using std::endl;
+using std::to_string;
+using std::vector;
+
+using EdgeId = int;
+
+struct Edge {
+  EdgeId id;
+  std::array<int, 2> nodes;
+
+  Edge(int start, int end, int _id) : id(_id) {
+    nodes[0] = start;
+    nodes[1] = end;
+  }
+
+  std::string to_json() const {
+    std::string res;
+    res = "{ \"id\": ";
+    res += to_string(id);
+    res += ", \"vertex_ids\": [";
+    res += to_string(nodes[0]);
+    res += ", ";
+    res += to_string(nodes[1]);
+    res += "] }";
+    return res;
+  }
+};
+
+// TODO check uniqueness of ID field
+struct Vertex {
+  int id = -1;
```

Comment:
В данном контексте `-1` - magic number, будет лучше создать константу
```cpp
constexpr int INVALID_ID = -1; 
```
```suggestion
  int id = INVALID_ID;
```